### PR TITLE
CLI: improve wallet creation

### LINKED ISSUE #2052: CLI: Refactor `InitParameters`?
## Description

- [x] does it make sense to have `localhost:8050` as the node default for the regular user?
- [x] does it make sense to have `4219/0/0/0` as the default bip path?
- [x] should we restrict certain coin types?
- [x] should we have individual fields on `InitParameters` for the bip path segments?

### PR DESCRIPTION
**Pull request by Alex6323**
# Description of change

This PR suggests the changes:
* allow to specify `alias` in the `InitParameters`, because that just seems right to allow that;
* removes the default bip path in `InitParameters`, because defaulting to a certain network could be confusing for some users;
* after the wallet has been successfully created, more information is printed and logged as a summary, because that not only helps us debugging from user provided log files, but also tells the user whether the wallet is how s/he wanted it to be created.

## Links to any relevant issues

Fixes #2052

### CODE REVIEW

File: cli/src/cli.rs [hunk 1]
```diff
@@ -67,12 +67,17 @@ pub struct InitParameters {
     /// Set the node to connect to with this wallet.
     #[arg(short, long, value_name = "URL", env = "NODE_URL", default_value = DEFAULT_NODE_URL)]
     pub node_url: String,
-    /// Set the BIP path, `4219/0/0/0` if not provided.
-    #[arg(short, long, value_parser = parse_bip_path, default_value = "4219/0/0/0")]
+    /// Set the BIP path. If not provided a bip path has to be provided interactively on first launch.
+    /// The expected format is: `a/b/c/d` where `a` = coin type, `b` = account index, `c` = address index, `d` =
+    /// internal.
```
Comment (DaughterOfMars) [inline cli/src/cli.rs]:
```suggestion
    /// The expected format is: `<coin_type>/<account_index>/<address_index>/<change>`.
```

Maybe this?

File: cli/src/cli.rs [hunk 2]
```diff
@@ -370,6 +350,16 @@ pub async fn new_wallet(cli: Cli) -> Result<Option<Wallet>, Error> {
     })
 }
 
+async fn print_create_wallet_summary(wallet: &Wallet, secret_manager_variant: &str) -> Result<(), Error> {
+    println_log_info!("Created new wallet with the following parameters:",);
+    println_log_info!("Secret manager: {secret_manager_variant}");
+    println_log_info!("Wallet address: {}", wallet.address().await);
+    println_log_info!("Wallet bip path: {:?}", wallet.bip_path().await);
+    println_log_info!("Wallet alias: {:?}", wallet.alias().await);
```
Comment (Thoralf-M) [inline cli/src/cli.rs]: No debug print needed?
```suggestion
    println_log_info!("Wallet alias: {}", wallet.alias().await);
```

Comment (Alex6323) [reply cli/src/cli.rs]: it's an `Option` what gets returned

Comment (Thoralf-M) [reply cli/src/cli.rs]: Ah true

File: cli/src/cli.rs [hunk 3]
```diff
@@ -400,23 +389,37 @@ pub async fn init_command(
...
+    let mut address = init_params.address.map(|s| Bech32Address::from_str(&s)).transpose()?;
+    if address.is_none() {
+        if get_decision("Do you want to assign an existing address to your new wallet?")? {
+            address.replace(get_address("Set wallet address").await?);
+        }
+    }
+
+    let mut bip_path = init_params.bip_path;
+    if bip_path.is_none() {
+        if get_decision("Do you want to assign a valid address bip path to your new wallet?")? {
```
Comment (Thoralf-M) [inline cli/src/cli.rs]: `valid address` sounds useless here or can I set a bip path for an invalid address? :trollface:

File: cli/src/cli.rs [hunk 4]
```diff
@@ -400,23 +389,37 @@ pub async fn init_command(
...
+            bip_path.replace(
+                get_bip_path("Set address bip path (format: a/b/c/d)")
+                    .await
+                    .expect("todo"),
```
Comment (Thoralf-M) [inline cli/src/cli.rs]: todo

File: cli/src/helper.rs [hunk 5]
```diff
@@ -61,6 +62,53 @@ pub async fn get_alias(prompt: &str) -> Result<String, Error> {
     }
 }
 
+pub async fn get_address(prompt: &str) -> Result<Bech32Address, Error> {
+    loop {
+        let input = Input::<String>::new().with_prompt(prompt).interact_text()?;
+        if input.is_empty() || !input.is_ascii() {
+            println_log_error!("Invalid input, please choose a non-empty address consisting of ASCII characters.");
+        } else {
+            return Ok(Bech32Address::from_str(&input)?);
```
Comment (Thoralf-M) [inline cli/src/helper.rs:71]: todo

File: cli/src/helper.rs [hunk 6]
```diff
@@ -61,6 +62,53 @@ pub async fn get_alias(prompt: &str) -> Result<String, Error> {
...
+    }
+}
+
+pub async fn get_bip_path(prompt: &str) -> Result<Bip44, Error> {
+    loop {
+        let input = Input::<String>::new().with_prompt(prompt).interact_text()?;
+        if input.is_empty() || !input.is_ascii() {
+            println_log_error!("Invalid input, please choose a non-empty address consisting of ASCII characters.");
+        } else {
+            return Ok(parse_bip_path(&input).expect("todo"));
```
Comment (Thoralf-M) [inline cli/src/helper.rs]: todo

File: cli/src/cli.rs (hunk 3 above)
Comment (Alex6323) [reply cli/src/cli.rs]: it was meant to be read as "valid address-bip-path", but I gotcha :smiley:

File: cli/src/helper.rs [hunk 7]
```diff
@@ -61,6 +62,53 @@ pub async fn get_alias(prompt: &str) -> Result<String, Error> {
...
+            return Ok(Bech32Address::from_str(&input)?);
+        }
+    }
+}
+
+pub async fn get_bip_path(prompt: &str) -> Result<Bip44, Error> {
+    loop {
+        let input = Input::<String>::new().with_prompt(prompt).interact_text()?;
+        if input.is_empty() || !input.is_ascii() {
+            println_log_error!("Invalid input, please choose a non-empty address consisting of ASCII characters.");
```
Comment (Thoralf-M) [inline cli/src/helper.rs]: Why address?

Comment (Alex6323) [reply cli/src/helper.rs]: well, you know ... there's this incredibly efficient but error-prone coding style ... what was it called? uhm ... co ... cop ... copy & paste or somethin :trollface:

File: cli/src/cli.rs (hunk 3 above)
Comment (Alex6323) [reply cli/src/cli.rs]: Changed

File: cli/src/cli.rs (hunk 4 above)
Comment (Alex6323) [reply cli/src/cli.rs]: Thanks ... fixed

File: cli/src/helper.rs (hunk 6 above)
Comment (Alex6323) [reply cli/src/helper.rs]: Thanks, fixed

File: cli/src/helper.rs (hunk 7 above)
Comment (Alex6323) [reply cli/src/helper.rs]: Thanks, fixed

File: cli/src/helper.rs [hunk 8]
```diff
@@ -54,13 +55,62 @@ pub async fn get_alias(prompt: &str) -> Result<String, Error> {
...
+        }
+    }
+}
+
+pub async fn get_bip_path(prompt: &str) -> Result<Bip44, Error> {
+    loop {
+        let input = Input::<String>::new().with_prompt(prompt).interact_text()?;
+        if input.is_empty() || !input.is_ascii() {
+            println_log_error!(
+                "Invalid input, please enter a valid bip path (<coin_type>/<account_index>/<change_address>/<address_index>."
```
Comment (DaughterOfMars) [inline cli/src/helper.rs]:
```suggestion
                "Invalid input, please enter a valid bip path (<coin_type>/<account_index>/<change_address>/<address_index>)."
```

File: cli/src/cli.rs [hunk 9]
```diff
@@ -400,23 +390,39 @@ pub async fn init_command(
...
+        } else {
+            forced = true;
+        }
+    }
+
+    let mut bip_path = init_params.bip_path;
+    if bip_path.is_none() {
+        if forced || get_decision("Do you want to set the bip path of the new wallet?")? {
+            bip_path.replace(
+                get_bip_path("Set bip path (format=<coin_type>/<account_index>/<change_address>/<address_index>)")
```
Comment (DaughterOfMars) [inline cli/src/cli.rs]:
```suggestion
                get_bip_path("Set bip path (<coin_type>/<account_index>/<change_address>/<address_index>)")
```

File: cli/src/helper.rs [hunk 10]
```diff
@@ -54,13 +55,62 @@ pub async fn get_alias(prompt: &str) -> Result<String, Error> {
...
+            Ok(value) => bip_path_enc.push(value),
+            Err(_) => {
+                return Err(format!("cannot parse BIP path: {p}"));
+            }
+        }
+    }
+
+    if bip_path_enc.len() != 4 {
+        return Err(
+            "invalid BIP path format. Expected: `coin_type/account_index/change_address/address_index`".to_string(),
```
Comment (DaughterOfMars) [inline cli/src/helper.rs]:
```suggestion
            "invalid BIP path format. Expected: `<coin_type>/<account_index>/<change_address>/<address_index>`".to_string(),
```

**Review by DaughterOfMars — DISMISSED** [review]

File: cli/src/cli.rs [hunk 11]
```diff
@@ -400,23 +390,38 @@ pub async fn init_command(
...
+        if get_decision("Do you want to set the address of the new wallet?")? {
+            address.replace(get_address("Set wallet address").await?);
+        } else {
+            forced = true;
+        }
+    }
+
+    let mut bip_path = init_params.bip_path;
+    if bip_path.is_none() {
+        if forced || get_decision("Do you want to set the bip path of the new wallet?")? {
```
Comment (Thoralf-M) [inline cli/src/cli.rs:405]:
```
Do you want to set the address of the new wallet?: n
Set bip path (<coin_type>/<account_index>/<change_address>/<address_index>): n
cannot parse BIP path: n
```
That's a bit scam
Also I think we should at least display the default bip path for IOTA (4218/0/0/0) and Shimmer (4219/0/0/0) or add an option for the user to select the default or write a custom one (similar to how one can select to generate a mnemonic or provide a custom one)

File: cli/src/cli.rs [hunk 12]
```diff
@@ -400,23 +390,38 @@ pub async fn init_command(
...
+            bip_path.replace(
+                get_bip_path("Set bip path (<coin_type>/<account_index>/<change_address>/<address_index>)").await?,
```
Comment (Thoralf-M) [inline cli/src/cli.rs]: I think we should at least display the default bip path for IOTA (4218/0/0/0) and Shimmer (4219/0/0/0) or add an option for the user to select the default or write a custom one (similar to how one can select to generate a mnemonic or provide a custom one)

Comment (Alex6323) [reply cli/src/cli.rs]: I agree. I will add that.

Comment (Alex6323) [reply cli/src/cli.rs]: Implemented your suggestion.

File: cli/src/helper.rs [hunk 13]
```diff
@@ -50,13 +55,74 @@ pub fn get_decision(prompt: &str) -> Result<bool, Error> {
...
+            Err(err) => {
+                println_log_error!("Invalid input, please enter a valid Bech32 address: {err}");
+            }
+        }
+    }
+}
+
+pub fn parse_bip_path(input: &str) -> Result<Bip44, Error> {
+    if input.is_empty() || !input.is_ascii() {
+        return Err(eyre!(
```
Comment (DaughterOfMars) [inline cli/src/helper.rs]: `bail!`

File: cli/src/helper.rs [hunk 14]
```diff
@@ -50,13 +55,74 @@ pub fn get_decision(prompt: &str) -> Result<bool, Error> {
...
+        match segment.parse::<u32>() {
+            Ok(s) => segments.push(s),
+            Err(err) => {
+                return Err(eyre!("invalid BIP path segment. {i}/`{segment}`: {err}"));
+            }
+        }
+    }
+
+    if segments.len() != 4 {
+        return Err(eyre!(
```
Comment (DaughterOfMars) [inline cli/src/helper.rs]: `bail!`

File: cli/src/helper.rs [hunk 15]
```diff
@@ -50,13 +55,74 @@ pub fn get_decision(prompt: &str) -> Result<bool, Error> {
...
+            "invalid BIP path format. Expected: `<coin_type>/<account_index>/<change_address>/<address_index>`"
+        ));
+    }
+
+    let mut segments = Vec::with_capacity(4);
+    for (i, segment) in input.split_terminator('/').map(|p| p.trim()).enumerate() {
+        match segment.parse::<u32>() {
+            Ok(s) => segments.push(s),
+            Err(err) => {
+                return Err(eyre!("invalid BIP path segment. {i}/`{segment}`: {err}"));
```
Comment (DaughterOfMars) [inline cli/src/helper.rs]: `bail!`

**Review by DaughterOfMars — DISMISSED** [review]

File: cli/src/cli.rs [hunk 16]
```diff
@@ -395,38 +378,54 @@ pub async fn change_password_command(wallet: &Wallet, current_password: Password
...
-        None
-    };
+pub async fn init_command(storage_path: &Path, init_params: InitParameters) -> Result<Wallet, Error> {
+    let secret_manager = create_secret_manager(&init_params).await?;
+    let secret_manager_variant = secret_manager.to_string();
+
+    let mut address = init_params.address.map(|s| Bech32Address::from_str(&s)).transpose()?;
+    let mut forced = false;
+    if address.is_none() {
+        if enter_decision("Do you want to set the address of the new wallet?")? {
```
Comment (Thoralf-M) [inline cli/src/cli.rs]: `enter_decision()` has default `yes` which should be false for the address I'd say

File: cli/src/cli.rs [hunk 17]
```diff
@@ -67,12 +68,16 @@ pub struct InitParameters {
...
-    #[arg(short, long, value_parser = parse_bip_path, default_value = "4219/0/0/0")]
+    /// Set the BIP path. If not provided a bip path has to be provided interactively on first launch.
+    /// The expected format is: `<coin_type>/<account_index>/<change_address>/<address_index>`.
+    #[arg(short, long, value_parser = parse_bip_path)]
     pub bip_path: Option<Bip44>,
     /// Set the Bech32-encoded wallet address.
     #[arg(short, long)]
     pub address: Option<String>,
+    /// Set the wallet alias name.
+    #[arg(short = 'l', long)]
```
Comment (thibault-martinez) [inline cli/src/cli.rs:79]: Why `l` ?

Comment (Alex6323) [reply cli/src/cli.rs:79]: because `a` is already used

Comment (thibault-martinez) [reply cli/src/cli.rs:79]: Ah yeah sure, was looking for an explicit `a`. Not fully awake yet apparently, thanks

File: cli/src/cli.rs (hunk 16 above)
Comment (Alex6323) [reply cli/src/cli.rs]: Changed

**Review by Thoralf-M — APPROVED** [review]

**Review by DaughterOfMars — APPROVED** [review]

**Pull request merged**